src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/internal/jaxrs/java/DefaultJaxRsInfoProvider.java

```diff
@@ -0,0 +1,163 @@
+import com.redhat.devtools.intellij.lsp4mp4ij.psi.core.utils.PsiTypeUtils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static com.redhat.devtools.intellij.lsp4mp4ij.psi.core.jaxrs.JaxRsConstants.JAKARTA_WS_RS_PATH_ANNOTATION;
+import static com.redhat.devtools.intellij.lsp4mp4ij.psi.core.jaxrs.JaxRsConstants.JAVAX_WS_RS_PATH_ANNOTATION;
+import static com.redhat.devtools.intellij.lsp4mp4ij.psi.core.jaxrs.JaxRsUtils.getJaxRsPathValue;
+import static com.redhat.devtools.intellij.lsp4mp4ij.psi.core.jaxrs.JaxRsUtils.isJaxRsRequestMethod;
+import static com.redhat.devtools.intellij.lsp4mp4ij.psi.core.utils.AnnotationUtils.hasAnnotation;
+import static com.redhat.devtools.intellij.lsp4mp4ij.psi.core.utils.PsiTypeUtils.overlaps;
+
+
+/**
+ * Locates JAX-RS methods in a project or class file using the default semantics.
+ */
+public class DefaultJaxRsInfoProvider implements IJaxRsInfoProvider {
+
+	private static final Logger LOGGER = Logger.getLogger(DefaultJaxRsInfoProvider.class.getName());
+
+	@Override
+	public boolean canProvideJaxRsMethodInfoForClass(PsiFile typeRoot, Module javaProject, ProgressIndicator monitor) {
+		return PsiTypeUtils.findType(javaProject, JAVAX_WS_RS_PATH_ANNOTATION) == null
+				|| PsiTypeUtils.findType(javaProject, JAKARTA_WS_RS_PATH_ANNOTATION) == null;
+	}
+
+	@Override
+	public Set<PsiClass> getAllJaxRsClasses(Module javaProject, ProgressIndicator monitor) {
+		// TODO: implement when LSP4IJ will support workspace symbols
+		return Collections.emptySet();
+	}
+
+	@Override
+	public List<JaxRsMethodInfo> getJaxRsMethodInfo(PsiFile typeRoot, JaxRsContext jaxrsContext, IPsiUtils utils,
+													ProgressIndicator monitor) {
+		List<JaxRsMethodInfo> methodInfos = new ArrayList<>();
+		try {
+			collectJaxRsMethodInfo(typeRoot.getChildren(), null, methodInfos, jaxrsContext, utils, monitor);
+		} catch (Exception e) {
+			LOGGER.log(Level.SEVERE, "while collecting JAX-RS method info using the default method", e);
+		}
+		return methodInfos;
+	}
+
+	private static void collectJaxRsMethodInfo(PsiElement[] elements, String rootPath,
+											   Collection<JaxRsMethodInfo> jaxRsMethodsInfo, JaxRsContext jaxrsContext, IPsiUtils utils,
+											   ProgressIndicator monitor) {
+		for (PsiElement element : elements) {
+			if (monitor.isCanceled()) {
+				return;
+			}
+			if (element instanceof PsiClass) {
+				PsiClass type = (PsiClass) element;
+				// Get value of JAX-RS @Path annotation from the class
+				String pathValue = getJaxRsPathValue(type);
+				if (pathValue != null) {
+					// Class is annotated with @Path
+					// Loop for each method annotated with @Path to generate
+					// URL code lens per
+					// method.
+					collectJaxRsMethodInfo(type.getChildren(), pathValue, jaxRsMethodsInfo, jaxrsContext, utils,

```

Comment:
how about using type.getMethods() or getAllMethods() instead? it should be more efficient

Reply:
I'm not sure that it will be more efficient and I would like to keep this code to looks like to JDT implementation.